modules/host/persist.go

```diff
@@ -1,80 +1,165 @@
+
+	// Statistics.
+	Profit types.Currency
 }
 
-func (h *Host) save() error {
-	sHost := savedHost{
-		SpaceRemaining: h.spaceRemaining,
-		FileCounter:    h.fileCounter,
-		Profit:         h.profit,
-		HostSettings:   h.HostSettings,
-		Obligations:    make([]contractObligation, 0, len(h.obligationsByID)),
-		SecretKey:      h.secretKey,
-		PublicKey:      h.publicKey,
-	}
+// getObligations returns a slice containing all of the contract obligations
+// currently being tracked by the host.
+func (h *Host) getObligations() []contractObligation {
+	cos := make([]contractObligation, 0, len(h.obligationsByID))
 	for _, ob := range h.obligationsByID {
-		// to avoid race conditions involving the obligation's mutex, copy it
-		// manually into a new object
-		obcopy := contractObligation{ID: ob.ID, FileContract: ob.FileContract, LastRevisionTxn: ob.LastRevisionTxn, Path: ob.Path}
-		sHost.Obligations = append(sHost.Obligations, obcopy)
+		cos = append(cos, contractObligation{
+			ID:              ob.ID,
+			FileContract:    ob.FileContract,
+			LastRevisionTxn: ob.LastRevisionTxn,
+			Path:            ob.Path,
+		})
 	}
-
-	return persist.SaveFile(persistMetadata, sHost, filepath.Join(h.persistDir, "settings.json"))
+	return cos
 }
 
-func (h *Host) load() error {
-	var sHost savedHost
-	err := persist.LoadFile(persistMetadata, &sHost, filepath.Join(h.persistDir, "settings.json"))
-	if err != nil {
-		return err
+// save stores all of the persist data to disk.
+func (h *Host) save() error {
+	p := persistence{
+		BlockHeight:  h.blockHeight,
+		RecentChange: h.recentChange,
+		PublicKey:    h.publicKey,
+		SecretKey:    h.secretKey,
+		Settings:     h.settings,
+
+		FileCounter: h.fileCounter,
+		Obligations: h.getObligations(),
+
+		Profit: h.profit,
 	}
+	return persist.SaveFile(persistMetadata, p, filepath.Join(h.persistDir, "settings.json"))
+}
 
-	h.spaceRemaining = sHost.HostSettings.TotalStorage
-	h.fileCounter = sHost.FileCounter
-	h.HostSettings = sHost.HostSettings
-	h.profit = sHost.Profit
-	// recreate maps
-	for i := range sHost.Obligations {
-		obligation := &sHost.Obligations[i] // both maps should use same pointer
+// loadObligations loads file contract obligations from the persistent file
+// into the host.
+func (h *Host) loadObligations(cos []contractObligation) {
+	// Clear the existing obligations maps.
+	for i := range cos {
+		obligation := &cos[i] // both maps should use same pointer
 		height := obligation.FileContract.WindowStart + StorageProofReorgDepth
+		// Sanity check - if the height is below the current height, then set
+		// the height to current height + 3. This makes sure that all file
+		// contracts will eventually be hit or garbage collected by the host,
+		// even if a bug means that they aren't acted upon at the right moment.
+		if build.DEBUG && height < h.blockHeight {
+			panic("host settings file is inconsistent")
+		} else if height < h.blockHeight {
+			height = h.blockHeight + 3
+		}
```

Comment:
I didn't want to do `+1` because I was afraid of off-by-one errors somewhere. It doesn't really hurt to wait an extra block, and if the consensus stuff is going to be passing more blocks down the pipe, I thought there might be more reason to wait. I'm not really sure why 3 exactly, but it's not code that's supposed to be hit ever anyway. I guess you could run into problems if someone never ran their node for more than 3 blocks at a time.